### ISSUE TITLE
TypeError: undefined is not an object (evaluating 'xhr.setRequestHeader')

### ISSUE THREAD

**Issue by dpaola2** [hsgubert/turbolinks-form#3]
Hello,

I cannot get turbolinks-form to work as instructed in the README. I see the same behavior before and after adding the gem and configuring my app. I do see this in the JS console after submitting the form:

![screenshot 2018-08-20 15 00 39](https://user-images.githubusercontent.com/150509/44369053-d375fd00-a489-11e8-99c5-e8dcc3579629.png)

I'm testing this in Turbolinks 5.1.1, rails 5.2.0.

My controller:

```ruby
class SessionsController < ApplicationController
  def new
    authorize :session, :new?
    background_image
  end

  def create
    authorize :session, :create?
    user = User.find_by(email: params[:session][:email].downcase)
    if user && user.authenticate(params[:session][:password])
      log_in(user)
      remember(user)
      redirect_to home_path
    else
      flash[:error] = "Invalid email/password combination."
      background_image
      render_form_with_errors
    end
  end

  def destroy
    log_out
    redirect_to root_path
  end
end
```

My application.js:

```
//= require rails-ujs
//= require activestorage
//= require turbolinks
//= require_tree .
//= require turbolinks-form

```

Thanks in advance for any assistance you can offer.

**Comment by keeganquinn** [comment]
I ran into the same symptoms when I first started using this gem. Switching from rails-ujs to jquery-ujs resolved the situation for me. It appears that rails-ujs does not implement AJAX event triggers in the same way as jquery-ujs and this problem occurs as a result when it is used.

IMO this issue might be best resolved with a simple documentation note saying that rails-ujs is not compatible. Internally, the code does already clearly refer to jquery-ujs and not rails-ujs.

HTH!

**Comment by dpaola2** [comment]
Thanks. I ended up going with a different solution and not using this gem.

**Comment by hsgubert** [comment]
@keeganquinn thanks for sharing your conclusion.

This gem was created for a legacy project that uses **jquery_ujs** indeed, and not **rails_ujs**. But it would be nice to make this gem compatible with rails_ujs so it can be used with rails 5 and up.

This gem uses only 3 events of jquery_ujs:

- ajax:error
- ajax:success
- ajax:beforeSend

As it seems, those same 3 events still exist on rails-ujs [current code reference](https://github.com/rails/rails/blob/master/actionview/app/assets/javascripts/rails-ujs/features/remote.coffee#L60)

It appears you have investigated this issue in the past, any ideias why the gem fails with rails-ujs? Or is it only a matter of removing jquery_ujs from the assets and all should work fine?

**Comment by hsgubert** [comment]
Humm, just found out the event handlers signature changed... 
[reference in rails guides](https://edgeguides.rubyonrails.org/working_with_javascript_in_rails.html#rails-ujs-event-handlers)

Perhaps I can make a JS code that works for both signatures...

**Comment by hsgubert** [comment]
Fixed it on v0.1.1.

Now the gem is compatible with both jquery-ujs and rails-ujs (but it is up to the developer to add one of them to the assets).

Also, the gem no longer requires JQuery. It was a very direct process to remove JQuery dependencies from the JS code.

**Comment by keeganquinn** [comment]
Just a quick note here, I'm seeing regressions after updating to 0.1.2. When clicking a remote link which does not have any turbolinks-form tag:

[Error] TypeError: null is not an object (evaluating 'currentTarget.matches')
	(anonymous function) (main-2f4cea9b13ff9e6be919.js:19574)
	dispatch (main-2f4cea9b13ff9e6be919.js:10279:104)
	trigger (main-2f4cea9b13ff9e6be919.js:13178)
	(anonymous function) (main-2f4cea9b13ff9e6be919.js:13244)
	each (main-2f4cea9b13ff9e6be919.js:6004)
	fire (main-2f4cea9b13ff9e6be919.js:5198)
	beforeSend (main-2f4cea9b13ff9e6be919.js:5275)
	ajax (main-2f4cea9b13ff9e6be919.js:13981)
	(anonymous function) (main-2f4cea9b13ff9e6be919.js:5561)
	dispatch (main-2f4cea9b13ff9e6be919.js:10279:104)

From a brief glance at the code it looks like your matches polyfill may not be working as you expect. Also it appears your new on() function still calls $ from jQuery, so that is still a dependency or no?

I'll have more time to dig into this later today, I'll probably fork and put together a new issue + PR if I don't hear back before then. Meanwhile, sorry to bump the closed issue.

Thanks for your work on this!

**Comment by dpaola2** [comment]
On a related note, I heard an interview with Sam Stephenson on a podcast at some point in the last 6 months that they're working on Turbolinks 6, which will supposedly have better form submission handling. Coupled with the fact that turbolinks-android is apparently deprecated, I'd bet money that the next version of Turbolinks is going to be gangbusters.

**Comment by keeganquinn** [comment]
One last follow-up to say the last TypeError I mentioned is fixed in 0.1.3. Thanks again!

**Comment by hsgubert** [comment]
@dpaola2 I'm eagerly waiting for the next version of Turbolinks too. I hope they finally solve the shortcoming related to submits on the new version...

@keeganquinn I now it is not a good practice but I ended up publishing some versions of the gem so I could run all the regression tests on my CI. After publishing 0.1.3 all tests are passing again. (I use poltergeist and phantomJS on my integration tests)
